main/src/ca/uwaterloo/flix/language/phase/Dependencies.scala

```diff
@@ -1,88 +1,719 @@
+    visitExp(catchRule.exp)
+  }
+
+  private def visitHandlerRule(handlerRule: TypedAst.HandlerRule)(implicit sctx: SharedContext): Unit = {
+    visitSymUse(handlerRule.op)
+    visitExp(handlerRule.exp)
+    handlerRule.fparams.foreach(visitFParam)
+  }
+
+  private def visitJvmMethod(method: TypedAst.JvmMethod)(implicit sctx: SharedContext): Unit = {
+    visitType(method.retTpe)
+    visitType(method.eff)
+    visitExp(method.exp)
+    method.fparams.foreach(visitFParam)
+  }
+
+  private def visitSelectChannelRule(rule: TypedAst.SelectChannelRule)(implicit sctx: SharedContext): Unit = {
+    visitBnd(rule.bnd)
+    visitExp(rule.chan)
+    visitExp(rule.exp)
+  }
+
+  private def visitParYieldFragment(fragment: TypedAst.ParYieldFragment)(implicit sctx: SharedContext): Unit = {
+    visitExp(fragment.exp)
+    visitPattern(fragment.pat)
+  }
+
+  private def visitConstrait(constraint: TypedAst.Constraint)(implicit sctx: SharedContext): Unit = {
+    visitHead(constraint.head)
+    constraint.cparams.foreach(visitContraintParam)
+    constraint.body.foreach(visitConstraintBody)
+  }
+
+  private def visitHead(head: TypedAst.Predicate.Head)(implicit sctx: SharedContext): Unit = head match {
+    case TypedAst.Predicate.Head.Atom(_, _, terms, tpe, _) => visitType(tpe)
+      terms.foreach(visitExp)
+  }
+
+  private def visitContraintParam(cp: TypedAst.ConstraintParam)(implicit sctx: SharedContext): Unit = {
+    visitType(cp.tpe)
+    visitBnd(cp.bnd)
+  }
+
+  private def visitConstraintBody(cb: Body)(implicit sctx: SharedContext): Unit = cb match {
+    case Body.Atom(_, _, _, _, terms, tpe, _) =>
+      visitType(tpe)
+      terms.foreach(visitPattern)
+    case Body.Functional(outBnds, exp, _) =>
+      visitExp(exp)
+      outBnds.foreach(visitBnd)
+    case Body.Guard(exp, _) => visitExp(exp)
+  }
+
+  private def visitPParam(pparam: TypedAst.PredicateParam)(implicit sctx: SharedContext): Unit =
+    visitType(pparam.tpe)
+
+  private def visitOp(Op: TypedAst.Op)(implicit sctx: SharedContext): Unit =
+    visitSpec(Op.spec)
+
+  private def visitCase(cas: TypedAst.Case)(implicit sctx: SharedContext): Unit = {
+    visitScheme(cas.sc)
+    cas.tpes.foreach(visitType)
+  }
+
+  private def visitStructField(structField: TypedAst.StructField)(implicit sctx: SharedContext): Unit =
+    visitType(structField.tpe)
+
+  private def visitAssocTypeSig(assoc: TypedAst.AssocTypeSig)(implicit sctx: SharedContext): Unit =
+    assoc.tpe.toList.foreach(visitType)
+
+  private def visitSig(sig: TypedAst.Sig)(implicit sctx: SharedContext): Unit = {
+    visitSpec(sig.spec)
+    sig.exp.toList.foreach(visitExp)
+  }
+
+  private def visitExp(exp: Expr)(implicit sctx: SharedContext): Unit = exp match {
+    case Expr.Cst(_, tpe, _) =>
+      visitType(tpe)
+
+    case Expr.Var(_, tpe, _) =>
+      visitType(tpe)
+
+    case Expr.Hole(_, tpe, eff, _) =>
+      visitType(tpe)
+      visitType(eff)
+
+    case Expr.HoleWithExp(exp, tpe, eff, _) =>
```

Comment:
Here you have shadowing-- i.e. you have two exp.

We always try to avoid shadowing. For this reason we would often call the argument to visitExp exp0